Modify rule S5487: Adapt to LaYC format (#1919)
SonarSource/rspec: rules/S5487/cfamily/rule.adoc

Reviewer: We also have

* [S5486](https://sonarsource.github.io/rspec/#/rspec/S5486) "pthread_mutex_t" should not be consecutively locked or unlocked twice
* [S5489](https://sonarsource.github.io/rspec/#/rspec/S5489) "pthread_mutex_t" should be unlocked in the reverse order they were locked

on the same topic.

I think they could benefit from a similar update, or at least be made consistent (e.g., the compliant code examples introduce similar classes).

WDYT? I appreciate this extends the scope of your initial effort.

Reviewer 2: since the opening of this PR, these two rules have been thoroughly updated.
Do you think consistency is still missing?

Reviewer: I made this comment a while ago so I'd need a bit of time to reload the context here. From a cursory look at the rules, I think there might still be some room to improve consistency, such as the definition of mutexes. (The [`shared_content/cfamily`](https://github.com/SonarSource/rspec/tree/master/shared_content/cfamily)
 directory may help to share data across rules.)

The other thing I noted is the order of solutions: I would introduce solutions using the latest C++ version first, in all rules.

There may be more aspects to consider. I leave that to the discretion of the PR's reviewer. :-)

Failing to properly initialize or destroy a `pthread` mutex can lead to undefined behavior.

== Why is this an issue?

_Mutexes_ are synchronization primitives that allow managing concurrency.

Their use requires following a well-defined life cycle.

> * _Mutexes_ need to be initialized (``++pthread_mutex_init++``) before being used. Once it is initialized, a _mutex_ is in an _unlocked_ state.

Author: I would not recommend that replacement. ``++ avoids asciidoc logic to run.
Asciidoc grammar is a bit complex and I don't expect everybody working in RSPEC to know them by heart.
In that specific case you are right that it doesn't change anything, but I think we should have a simple rule like "If you have anything else than [A-Za-z0-9 \.] use ``++".

> * _Mutexes_ need to be initialized (``++pthread_mutex_init++``) before being used. Once it is initialized, a _mutex_ is in an _unlocked_ state.

Reviewer 3: Interesting. Sure, then let's keep the \`++code++` style here.

* _Mutexes_ need to be destroyed (``++pthread_mutex_destroy++``) to free the associated internal resources. Only _unlocked_ _mutexes_ can be safely destroyed.

> Before the initialization or after destruction, a mutex is in an uninitialized state.

Reviewer 3: I no longer feel that this is necessary.

During this life cycle,
the following patterns should be avoided as they result in undefined behavior:

* trying to initialize an already initialized _mutex_
* trying to destroy an initialized _mutex_ that is in a _locked_ state
* trying to destroy an uninitialized _mutex_
* trying to lock an uninitialized _mutex_
* trying to unlock an uninitialized _mutex_

In {cpp}11 and higher, `std::mutex` is less error-prone and is supported by more platforms.

In {cpp}03, it is recommended to wrap mutex creation/destruction in an RAII class, as well as mutex lock/unlock. Those RAII classes will perform the right operations, even in the presence of exceptions.

=== What is the potential impact?

Failing to properly initialize or destroy a POSIX Thread Mutex leads to *undefined behavior*.

For programs that exercise undefined behavior, the compiler is no longer bound by the language specification.
The application may crash or, even worse, the application may appear to execute correctly while losing data or producing incorrect results.
In a multi-threaded context, additionally, the application may experience spurious deadlocks or data races.

== How to fix it

> 

Reviewer 3: This section should also briefly mention "lock-free" as an alternative (using atomics).

[source,c]
----
pthread_mutex_t mtx1;

void double_initialization(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_init(&mtx1, NULL); // Noncompliant: initializing an already initialized mutex
}

void destroy_locked(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_lock(&mtx1);
  pthread_mutex_destroy(&mtx1); // Noncompliant: destroying a locked mutex
}

void double_destruction(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_destroy(&mtx1);
  pthread_mutex_destroy(&mtx1); // Noncompliant: destroying an uninitialized mutex
}

void lock_destroyed(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_destroy(&mtx1);
  pthread_mutex_lock(&mtx1); // Noncompliant: locking an uninitialized mutex
}

void unlock_destroyed(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_destroy(&mtx1);
  pthread_mutex_unlock(&mtx1); // Noncompliant: unlocking an uninitialized mutex
}
----

==== Compliant solution

C solution:

[source,c]
----
pthread_mutex_t mtx1;

void destroy_initialized(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_destroy(&mtx1);
}

void use_and_destroy_initialized(void)
{
  pthread_mutex_init(&mtx1, NULL);

>   pthread_mutex_lock(&mtx1);
>   pthread_mutex_unlock(&mtx1);

Reviewer 3: Let's add a comment to make it crystal clear.

```suggestion
  pthread_mutex_lock(&mtx1);
  // critical code that processes a shared resource (e.g. memory)
  pthread_mutex_unlock(&mtx1);
```

  pthread_mutex_destroy(&mtx1);
}
----

For static mutex variables:

[source,c]
----
pthread_mutex_t mtx1 = PTHREAD_MUTEX_INITIALIZER;

void use_statically_initialized(void)
{
  pthread_mutex_lock(&mtx1); // Compliant: mutex is initialized statically

>   pthread_mutex_unlock(&mtx1);

Reviewer 3: Same here. I would add a short comment in between.

}
----

{cpp}11 and higher solution:

[source,cpp]
----
std::mutex mtx1;

void use_lock_guard(bool condition) {
  std::lock_guard<std::mutex> lock(mtx1);
  if (condition) {
    // Compliant: mutex will unlock automatically even in case of an exception
    throw std::invalid_argument("Expected false");
  }
  // Compliant: mutex will unlock automatically at the end of the scope
}
----

{cpp}03 solution:

[source,cpp]
----
pthread_mutex_t mtx1;

class Mutex {
public:
  Mutex(pthread_mutex_t* mtx) : pmtx(mtx) {

>     pthread_mutex_init(pmtx, NULL);

Reviewer 3: ```suggestion
    pthread_mutex_init(pmtx, /*mutex attribute=*/NULL);
```

  }
  ~Mutex() {
    pthread_mutex_destroy(pmtx);
  }
  pthread_mutex_t* pmtx;
private:
  // Forbid the copying of the mutex to avoid double-free
  Mutex(Mutex const& other);
  Mutex& operator=(Mutex const& other);
};

struct LockGuard {
  LockGuard(Mutex &m) : mtx(m) {
    pthread_mutex_lock(mtx.pmtx);
  }
  ~LockGuard() {
    pthread_mutex_unlock(mtx.pmtx);
  }
  Mutex &mtx;
}

void destroy_initialized()
{

>   Mutex m(&mtx1);
>   // Compliant: mtx1 will be properly initialized and destroyed
> }

Reviewer 3: ```suggestion
  Mutex m(&mtx); // Compliant: mtx will be properly initialized
  // ... and destroyed at the end of this scope.
}
```

void use_and_destroy_initialized()
{
  Mutex m(&mtx1);
  {
    LockGuard lock(m);
    // Compliant: mutex will unlock automatically at the end of the scope
  }
  // Compliant: mtx1 will be destroyed properly
}
----

== Resources

> === Related rules

Reviewer 3: Please adjust the ordering: `Documentation` first, then
`Conference presentations`:
(This is a great talk that covers everything one needs to know concerning concurrency, locks, etc.)
```
=== Conference presentations

* CppCon 2020 - https://youtu.be/A7sVFJLJM-A?si=v76jhmv3XnHExZYU[An Introduction to Multithreading in C++20]
```
and lastly `Related rules`

* S5486 enforces the proper locking and unlocking of `pthread` mutexes.
* S5489 enforces unlocking multiple held `pthread` mutexes in reverse order.

=== Documentation

* The Open Group - https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_init.html[``++pthread_mutex_init++``, ``++pthread_mutex_destroy++``]

> * {cpp} Reference - https://en.cppreference.com/w/cpp/thread/lock_guard[`std::lock_guard`]

Reviewer 3: I would also add a dedicated link to `std::mutex` to suggest the use of C++, if available (to the developers).

> * {cpp} Reference - https://en.cppreference.com/w/cpp/thread/lock_guard[`std::lock_guard`]

Reviewer 2: ```suggestion
* {cpp} reference - https://en.cppreference.com/w/cpp/thread/lock_guard[`std::lock_guard`]
```

* trying to destroy an initialized _mutex_ that is in a _locked_ state
* trying to destroy an uninitialized _mutex_
* trying to lock an uninitialized _mutex_
* trying to unlock an uninitialized _mutex_

In {cpp}03, it is recommended to wrap mutex creation/destruction in an RAII class, as well as mutex lock/unlock. Those RAII classes will perform the right operations, even in the presence of exceptions.

In {cpp}11 and higher, `std::mutex` is less error-prone and is supported by more platforms.

== How to fix it

=== Code examples

==== Noncompliant code example

[source,cpp]
----
pthread_mutex_t mtx1;

void bad1(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_init(&mtx1, NULL);
}

void bad2(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_lock(&mtx1);
  pthread_mutex_destroy(&mtx1);
}

void bad3(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_destroy(&mtx1);
  pthread_mutex_destroy(&mtx1);
}

void bad4(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_destroy(&mtx1);
  pthread_mutex_lock(&mtx1);
}

void bad5(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_destroy(&mtx1);
  pthread_mutex_unlock(&mtx1);
}
----

> ==== Compliant solution

Reviewer 2: swapped ea0351fd11

C solution:

[source,c]
----
pthread_mutex_t mtx1;

> void ok1(void)

Reviewer 2: I went for "destroy_initialized" and "use_and_destroy_initialized", you can check it out in ea0351fd11

{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_destroy(&mtx1);
}

void ok2(void)
{
  pthread_mutex_init(&mtx1, NULL);
  pthread_mutex_lock(&mtx1);
  pthread_mutex_unlock(&mtx1);
  pthread_mutex_destroy(&mtx1);
}
----

{cpp}03 solution:

[source,cpp]
----
pthread_mutex_t mtx1;

> struct Mutex {

Reviewer 2: Movable is not a concern for C++03. I put the copy constructor into the `private` section ea0351fd11

  Mutex(pthread_mutex_t* mtx) : pmtx(mtx) {
    pthread_mutex_init(pmtx, NULL);
  }
  ~Mutex() {
    pthread_mutex_destroy(pmtx);
  }
  pthread_mutex_t* pmtx;
};

> struct Lock {

Reviewer: WDYT of using `LockGuard` to make a parallel with the C++11 version?
